### PR TITLE
Many-Particle Models

### PR DESCRIPTION
**Pull request by tobykirk**
# Description

This branch adds the Many-Particle Model or MPM, a full lithium-ion battery model including a distribution of particle sizes in each electrode. The particle size is implemented as an additional "dimension", where the user specifies the continuous particle-size distribution which is then discretized. The model, it's features and how to use it are detailed in the notebook `examples/notebooks/MPM.ipynb`.

Included is a new submodel module `particle.size_distribution` with 2 submodels `FickianSingleSizeDistribution` and `FastSingleSizeDistribution` (and a base distribution model) used by the MPM. These are the "X-averaged" ones, with submodels incorporating x-variation (`FickianManySizeDistributions`, etc.) to be merged later (pending the ability to have 4 domains). Changes to the existing code are currently quite minimal (probably to a fault!), mainly adding functionality where possible.

The branch includes all the base functionality for the new "particle size" domain/dimension, including additions to

- standard spatial variables (`R_n` and `R_p`)
- standard battery geometry
- lithium ion parameter class (e.g. area-weighted size distributions `f_a_dist_n`, `f_a_dist_p`)
- `BaseInterface` and `BaseKinetics` (where special treatment of variables that depend on "particle size" is needed) 
- `ProcessedVariable` 
- `QuickPlot`
- an `R_average()` operator, for convenient particle-size averaging

Not currently compatible with the degradation models, i.e., SEI, particle cracking/swelling, lithium plating - they are not allowed as options.

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: tests/unit/test_models/test_submodels/test_electrode/test_ohm/test_leading_size_distribution.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#
+# Test leading-order ohm submodel
```
Comment (tinosulzer) [inline tests/unit/test_models/test_submodels/test_electrode/test_ohm/test_leading_size_distribution.py:2]: I don't think these single-submodel tests are necessary any more. In practice they don't really help to catch bugs and are just a pain to maintain (changing which variables are input)

File: pybamm/models/submodels/particle/size_distribution/fickian_single_distribution.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
+#
+# Class for a single particle-size distribution representing an
+# electrode, with Fickian diffusion within each particle
```
Comment (tinosulzer) [inline pybamm/models/submodels/particle/size_distribution/fickian_single_distribution.py:3]: let's leave it for now but I wonder if this class could be merged with the no-size-distribution classes to reduce code duplication

**Review by tinosulzer — CHANGES_REQUESTED** [review]
Thanks @tobykirk , great to see this! Eventually we will want to make every variable live on the size distribution domain (which would be 0D in most cases) but I think this is fine on a first step. See some comments, mainly aimed at reducing code duplication

File: docs/source/expression_tree/unary_operator.rst [hunk 3]
```diff
@@ -81,6 +81,8 @@ Unary Operators
 
 .. autofunction:: pybamm.r_average
 
+.. autofunction:: pybamm.R_average
```
[the remark this answers is not in the document]
Comment (tobykirk) [reply docs/source/expression_tree/unary_operator.rst:84]: I agree, I've changed it to `size_average`

File: pybamm/expression_tree/unary_operators.py [hunk 4]
```diff
@@ -1393,6 +1399,67 @@ def r_average(symbol):
         return Integral(symbol, r) / Integral(v, r)
 
 
+def R_average(symbol, param):
```
[the remark this answers is not in the document]
Comment (tobykirk) [reply pybamm/expression_tree/unary_operators.py:1402]: Done

File: pybamm/models/full_battery_models/lithium_ion/mpm.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
+#
+# Many-Particle Model (MPM)
+#
+import pybamm
+from .base_lithium_ion_model import BaseModel
+
+
+class MPM(BaseModel):
```
[the remark this answers is not in the document]
Comment (tobykirk) [reply pybamm/models/full_battery_models/lithium_ion/mpm.py:8]: Excellent, I didn't think of this. I've now changed it to inherit from `SPM`. Annoyingly, one issue is that it can't use the `__init__` in `SPM` since the `MPM` needs to overwrite the citation `pybamm.citations.register("Marquis2019")`. Easily changed though, if you have a suggestion there.

File: pybamm/models/full_battery_models/lithium_ion/mpm.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+    def set_sei_submodel(self):
+
+        # negative electrode SEI
+        self.submodels["negative sei"] = pybamm.sei.NoSEI(self.param, "Negative")
+
+        # positive electrode
+        self.submodels["positive sei"] = pybamm.sei.NoSEI(self.param, "Positive")
+
+    @property
+    def default_parameter_values(self):
```
[the remark this answers is not in the document]
Comment (tobykirk) [reply pybamm/models/full_battery_models/lithium_ion/mpm.py:201]: Good idea, I've added a new file `size_distribution_parameters.py` in the `parameters` folder that does this. I've also made a separate `lognormal` function (but maybe that is better placed elsewhere?)

File: pybamm/parameters/lithium_ion_parameters.py [hunk 7]
```diff
@@ -331,6 +331,18 @@ def _set_dimensional_parameters(self):
...
+        self.R_min_p_dim = pybamm.Parameter("Positive minimum particle radius [m]")
+        self.R_max_n_dim = pybamm.Parameter("Negative maximum particle radius [m]")
+        self.R_max_p_dim = pybamm.Parameter("Positive maximum particle radius [m]")
+        self.sd_a_n_dim = pybamm.Parameter(
+            "Negative area-weighted particle-size standard deviation [m]"
+        )
+        self.sd_a_p_dim = pybamm.Parameter(
+            "Positive area-weighted particle-size standard deviation [m]"
+        )
+
```
[the remark this answers is not in the document]
Comment (tobykirk) [reply pybamm/parameters/lithium_ion_parameters.py:345]: Would be ideal, but the nondimensionalisation of the size distribution parameters uses the typical particle radius (`R_n_typ` etc) which are defined from the "Negative particle radius [m]" etc function parameters, defined in `lithium_ion_parameters.py`. Not sure if that can be moved - I can try it though

File: pybamm/models/submodels/interface/base_interface.py [hunk 8]
```diff
@@ -395,7 +444,13 @@ def _get_standard_exchange_current_variables(self, j0):
         elif self.domain == "Positive":
             j_scale = i_typ / (self.param.a_p_typ * L_x)
 
-        # Average, and broadcast if necessary
+        # If j0 depends on particle size R then must R-average to get standard
+        # output exchange current density
+        if j0.domain == [self.domain.lower() + " particle size"]:
+            # R-average
+            j0 = pybamm.R_average(j0, self.param)
```
[the remark this answers is not in the document]
Comment (tobykirk) [reply pybamm/models/submodels/interface/base_interface.py:451]: Yeah, I think this was a relic from before R_average was defined. I've removed the if statements. Also, many of the size-dependent variables (`j0`, `eta_r`, `ocp`) were not being output, only the size-averaged versions, but I've now added extra methods `_get_standard_size_distribution_ocp_variables` etc here to do so.

File: pybamm/models/submodels/interface/base_interface.py [hunk 9]
```diff
@@ -637,3 +710,74 @@ def _get_standard_ocp_variables(self, ocp, dUdT):
...
+        )
+
+        # current density that depends on particle size R
+        j_distribution = self._get_kinetics(j0, ne, eta_r, T)
+
+        # R-average
+        j = pybamm.R_average(j_distribution, self.param)
+        return j, j_distribution
+
+    def _get_standard_PSD_interfacial_current_variables(self, j_distribution):
```
[the remark this answers is not in the document]
Comment (tobykirk) [reply pybamm/models/submodels/interface/base_interface.py:737]: Changed to `_get_standard_size_distribution_current_variables`

File: pybamm/models/submodels/interface/base_interface.py [hunk 10]
```diff
@@ -637,3 +710,74 @@ def _get_standard_ocp_variables(self, ocp, dUdT):
             )
 
         return variables
+
+    def _get_PSD_current_densities(self, j0, ne, eta_r, T):
```
[the remark this answers is not in the document]
Comment (tobykirk) [reply pybamm/models/submodels/interface/base_interface.py:714]: Function removed as I've simplified how `j` is calculated in `base_kinetics`

File: pybamm/models/submodels/interface/inverse_kinetics/inverse_butler_volmer.py [hunk 11]
```diff
@@ -30,7 +30,10 @@ class InverseButlerVolmer(BaseInterface):
     def __init__(self, param, domain, reaction, options=None):
         super().__init__(param, domain, reaction)
         if options is None:
-            options = {"SEI film resistance": "none"}
+            options = {
+                "SEI film resistance": "none",
+                "particle size": "single"
```
[the remark this answers is not in the document]
Comment (tobykirk) [reply pybamm/models/submodels/interface/inverse_kinetics/inverse_butler_volmer.py:35]: Yeah, this is just to pass some unit tests (e.g. `test_inverse_butler_volmer.py`) that aren't passed the default options (same reason the `"SEI film resistance": "none"` was there, I think). Should I just move them/pass them in to the test instead?

File: pybamm/plotting/quick_plot.py [hunk 12]
```diff
@@ -342,14 +344,34 @@ def set_output_variables(self, output_variables, solutions):
...
+                        self.is_x_R[variable_tuple] = True
+                        self.is_R_r[variable_tuple] = False
+                    elif (
+                        first_spatial_var_name == "r" and second_spatial_var_name == "R"
+                    ):
+                        self.is_x_r[variable_tuple] = False
+                        self.is_y_z[variable_tuple] = False
+                        self.is_x_R[variable_tuple] = False
+                        self.is_R_r[variable_tuple] = True
                     else:
```
[the remark this answers is not in the document]
Comment (tobykirk) [reply pybamm/plotting/quick_plot.py:370]: I've now shortened these, lumping together the cases `is_x_r`, `is_x_R`, `is_R_r` together as they're treated the same way.

File: tests/unit/test_models/test_submodels/test_electrode/test_ohm/test_leading_size_distribution.py (hunk 1 above)
Comment (tobykirk) [reply tests/unit/test_models/test_submodels/test_electrode/test_ohm/test_leading_size_distribution.py:2]: This test is for an old version of a submodel and forgot to remove it. I see what you mean about them just being a pain (had to fix several other tests myself by adding variables/options)

File: tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_mpm.py [hunk 13]
```diff
@@ -0,0 +1,193 @@
+#
+# Tests for the lithium-ion MPM model
+#
```
[the remark this answers is not in the document]
Comment (tobykirk) [reply tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_mpm.py:3]: Ah I see. Do you have any suggestions of which to get rid of? It doesn't appear as though there are actual duplicates (as each test here has size distributions, which SPM doesn't have). Many of these tests are just testing that an error is raised though (as that submodel combination is not yet compatible), so perhaps I should remove those?

**Comment by tobykirk** [comment]
> Thanks @tobykirk , great to see this! Eventually we will want to make every variable live on the size distribution domain (which would be 0D in most cases) but I think this is fine on a first step. See some comments, mainly aimed at reducing code duplication

Thanks a lot @tinosulzer , I appreciate your comments and suggestions, they're really very helpful (and apologies for such a large pull request). I've made most of the requested changes and pushed the commits - see the replies to your comments.

File: tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_mpm.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (tobykirk) [reply tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_mpm.py:3]: Removed some of the tests, leaving a test for each thermal model and 1+1D current collectors (omitting 2+1D). The "not_implemented" tests can be removed too, if necessary.

File: tests/unit/test_models/test_full_battery_models/test_lithium_ion/test_mpm.py [hunk 14]
```diff
@@ -0,0 +1,247 @@
+#
+# Tests for the lithium-ion MPM model
```
[the remark this answers is not in the document]
Comment (tobykirk) [reply tests/unit/test_models/test_full_battery_models/test_lithium_ion/test_mpm.py:2]: Removed some of the tests, leaving a test for each thermal model and 1+1D current collectors (omitting 2+1D). The "not_implemented" tests can be removed too, if necessary.

File: pybamm/parameters/lithium_ion_parameters.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tinosulzer) [reply pybamm/parameters/lithium_ion_parameters.py:345]: we can make negative particle radius etc geometric parameters

File: pybamm/models/full_battery_models/lithium_ion/mpm.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tinosulzer) [reply pybamm/models/full_battery_models/lithium_ion/mpm.py]: could do an if statement in the SPM like
```python3
if self.__class__ != "MPM":
```

File: pybamm/models/submodels/interface/inverse_kinetics/inverse_butler_volmer.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (tinosulzer) [reply pybamm/models/submodels/interface/inverse_kinetics/inverse_butler_volmer.py]: yeah it's a bit clunky, pass them to the test instead - forgot the reason why we did this

File: tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_mpm.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (tinosulzer) [reply tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_mpm.py:3]: I think it's ok to just test the "MPM" part on its own with default options, and then hope that everything else just works. Same way as we don't test 1+1D with every single combination of options. If there is some combination that proves to be particularly tricky we can add tests.

File: pybamm/models/full_battery_models/lithium_ion/mpm.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tobykirk) [reply pybamm/models/full_battery_models/lithium_ion/mpm.py]: Great, added. I suppose in general you could add a method `set_citations` to the model, to be easily overwritten in any further subclass

File: pybamm/parameters/lithium_ion_parameters.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tobykirk) [reply pybamm/parameters/lithium_ion_parameters.py:345]: Great. I've made the particle radii and these size distribution params geometric parameters everything seems to work fine.

File: pybamm/expression_tree/unary_operators.py [hunk 15]
```diff
@@ -1393,6 +1399,67 @@ def r_average(symbol):
...
+    """convenience function for averaging over particle size R using the area-weighted
+    particle-size distribution.
+
+    Parameters
+    ----------
+    symbol : :class:`pybamm.Symbol`
+        The function to be averaged
+    param : :class:`pybamm.LithiumIonParameters`
```
[the remark this answers is not in the document]
Comment (tobykirk) [reply pybamm/expression_tree/unary_operators.py:1410]: Removed the `param` requirement, now that they are in `geometric_parameters`

File: pybamm/models/submodels/interface/inverse_kinetics/inverse_butler_volmer.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (tobykirk) [reply pybamm/models/submodels/interface/inverse_kinetics/inverse_butler_volmer.py]: Ok done!

File: tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_mpm.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (tobykirk) [reply tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_mpm.py:3]: Ah thanks, it is starting to make more sense. Testing all combinations of options is pretty unsustainable, so you add tests as the problems arise

File: pybamm/parameters/size_distribution_parameters.py [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+import pybamm
+import numpy as np
+
+
+def get_size_distribution_parameters(
+    param,
+    R_n_av=None,
+    R_p_av=None,
+    sd_n=None,
+    sd_p=None,
```
Comment (tinosulzer) [inline pybamm/parameters/size_distribution_parameters.py:15]: you could just put 0.3 here instead of None

**Review by tinosulzer — APPROVED** [review]
Thanks, looks great now!

**Comment by tinosulzer** [comment]
You can merge develop and push again to trigger the tests with the fix to new_copy

File: pybamm/models/submodels/particle/size_distribution/fickian_single_distribution.py [hunk 17]
```diff
@@ -106,18 +106,15 @@ def get_coupled_variables(self, variables):
...
         if self.domain == "Negative":
             N_s_xav_distribution = -self.param.D_n(
                 c_s_xav_distribution, T_k_xav
-            ) * pybamm.grad(c_s_xav_distribution) / R
+            ) * pybamm.grad(c_s_xav_distribution)
         elif self.domain == "Positive":
             N_s_xav_distribution = -self.param.D_p(
                 c_s_xav_distribution, T_k_xav
-            ) * pybamm.grad(c_s_xav_distribution) / R
+            ) * pybamm.grad(c_s_xav_distribution)
```
Comment (tinosulzer) [inline pybamm/models/submodels/particle/size_distribution/fickian_single_distribution.py:117]: why this change?

Comment (tobykirk) [reply pybamm/models/submodels/particle/size_distribution/fickian_single_distribution.py:117]: This is just to make the definition of the nondimensional flux consistent with `FickianManyParticles` (where you can have a size distribution in x), and also the scaling used for the MP-DFN in our recent JES paper. The `rhs` is unchanged, as the factor of `/ R` is just moved to lines 166 and 172, turning an `/ R` into `/ R ** 2`. Should make it easier to combine the "distribution" and "no distribution" submodels later on.

Meant to make this change much earlier, but seems I forgot. Sorry for the confusion here!

**Comment by tobykirk** [comment]
To describe the last few commits: I've added a few more tests to increase the coverage, particularly for `processed_variable.py` (testing variables that depend on particle size), `broadcasts.py` (to/from size domain) and `mpm.py` (testing the default parameter values and necessary options). 

For the `processed_variable` tests, I needed to modify `get_mesh_for_testing()` and `get_discretization_for_testing()`. (Maybe the tests here get repetitive and could be shortened, though?)

**Comment by tobykirk** [comment]
Not passing the total coverage check as I think codecov is comparing to an older commit (f83f0a2) rather than e.g. 396c9ef ?

**Comment by tinosulzer** [comment]
Yeah we could definitely reduce some duplication in the processed variable tests

**Comment by tobykirk** [comment]
> Yeah we could definitely reduce some duplication in the processed variable tests

Had a go at reducing the duplication by moving code into a method `process_and_check_2D_variable` (x-r, r-R, R-x, x-z, R-z variables are all tested almost identically)

**Comment by tinosulzer** [comment]
@all-contributors add @tobykirk for ideas, code, tests
